src/Storages/FileLog/StorageFileLog.cpp

```diff
@@ -0,0 +1,764 @@
+        }
+        if (!tryReadIntText(last_written_pos, buf))
+        {
+            throw Exception(ErrorCodes::READ_META_FILE_FAILED, "Read meta file {} failed.", dir_entry.path().c_str());
+        }
+
+        meta.file_name = dir_entry.path().filename();
+        meta.last_writen_position = last_written_pos;
+
+        /// May have last open end in meta file
+        if (checkChar('\n', buf))
+        {
+            if (!tryReadIntText(meta.last_open_end, buf))
+            {
+                throw Exception(ErrorCodes::READ_META_FILE_FAILED, "Read meta file {} failed.", dir_entry.path().c_str());
+            }
+        }
+
+        file_infos.meta_by_inode.emplace(inode, meta);
+    }
+}
+
+UInt64 StorageFileLog::getInode(const String & file_name)
+{
+    struct stat file_stat;
+    if (stat(file_name.c_str(), &file_stat))
+    {
+        throw Exception(ErrorCodes::CANNOT_GET_FILE_STAT, "Can not get stat info of file {}", file_name);
+    }
+    return file_stat.st_ino;
+}
+
+Pipe StorageFileLog::read(
+    const Names & column_names,
+    const StorageMetadataPtr & metadata_snapshot,
+    SelectQueryInfo & /* query_info */,
+    ContextPtr local_context,
+    QueryProcessingStage::Enum /* processed_stage */,
+    size_t /* max_block_size */,
+    unsigned /* num_streams */)
+{
+    std::lock_guard<std::mutex> lock(status_mutex);
+
+    updateFileInfos();
+
+    /// No files to parse
+    if (file_infos.file_names.empty())
+    {
+        return Pipe{};
+    }
+
+    auto modified_context = Context::createCopy(local_context);
+
+    auto max_streams_number = std::min<UInt64>(filelog_settings->filelog_max_threads, file_infos.file_names.size());
+
+    openFilesAndSetPos();
+    serialize(true);
+
+    Pipes pipes;
+    pipes.reserve(max_streams_number);
+    for (size_t stream_number = 0; stream_number < max_streams_number; ++stream_number)
+    {
+        Pipe pipe(std::make_shared<FileLogSource>(
+            *this, metadata_snapshot, modified_context, getMaxBlockSize(), getPollTimeoutMillisecond(), stream_number, max_streams_number));
+
+        auto convert_actions_dag = ActionsDAG::makeConvertingActions(
+            pipe.getHeader().getColumnsWithTypeAndName(),
+            metadata_snapshot->getSampleBlockForColumns(column_names, getVirtuals(), getStorageID()).getColumnsWithTypeAndName(),
+            ActionsDAG::MatchColumnsMode::Name);
+
+        auto actions = std::make_shared<ExpressionActions>(
+            convert_actions_dag, ExpressionActionsSettings::fromContext(getContext(), CompileExpressions::yes));
+
+        pipe.addSimpleTransform([&](const Block & stream_header) { return std::make_shared<ExpressionTransform>(stream_header, actions); });
+        pipes.emplace_back(std::move(pipe));
+    }
+
+    return Pipe::unitePipes(std::move(pipes));
+}
+
+void StorageFileLog::drop()
+{
+    if (std::filesystem::exists(root_meta_path))
+        std::filesystem::remove_all(root_meta_path);
```

Comment:
There is a problem for drop table, DatabaseAtomic does not call it when drop table.

Reply:
Yes, it will call it, but with some delay in the background. 
Actually I think we need a method for `IStorage` according to which we will decide whether we might need to call `IStorage::drop()` immediately and not in the background because there is no other method in `IStorage` which is called only when the table is dropped.
The same problem was also with rabbitmq table engine and it was solved with a really bad hack. So if there are table engines which need to call `drop()` at once (and for StorageFIleLog I think it is important), then I think we should allow it.